<!DOCTYPE html>
<html lang=zh>
<head>
    <!-- so meta -->
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="HandheldFriendly" content="True">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=5" />
    <meta name="description" content="HOMER 基本功能HOMER(Hypergeometric Optimization of Motif EnRichment)用于Motif identification和下一代测序分析，是一款很方便的软件。 *最不方便的地方在于安装… 笔者花了整整两天才部署完，有很多小坑。 一.homer在cygwin的安装网路上已有很多现成的安装教程资料了.这边仅记录一些在cygwin中安装的问题和解决方法">
<meta property="og:type" content="article">
<meta property="og:title" content="HOMER Silly Insturction">
<meta property="og:url" content="http://example.com/2023/06/19/HOMER-Silly-Instruction/index.html">
<meta property="og:site_name" content="Demo Draft">
<meta property="og:description" content="HOMER 基本功能HOMER(Hypergeometric Optimization of Motif EnRichment)用于Motif identification和下一代测序分析，是一款很方便的软件。 *最不方便的地方在于安装… 笔者花了整整两天才部署完，有很多小坑。 一.homer在cygwin的安装网路上已有很多现成的安装教程资料了.这边仅记录一些在cygwin中安装的问题和解决方法">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="s:\BlogProject\blog\source_posts\HOMER-Silly-Tutorials_pics\motifs.tssHistogram.png">
<meta property="article:published_time" content="2023-06-19T06:11:39.000Z">
<meta property="article:modified_time" content="2023-06-19T13:09:32.514Z">
<meta property="article:author" content="r3zgh0st">
<meta property="article:tag" content="test">
<meta property="article:tag" content="BIOINFO">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="s:\BlogProject\blog\source_posts\HOMER-Silly-Tutorials_pics\motifs.tssHistogram.png">
    
    
      
        
          <link rel="shortcut icon" href="/images/favicon.ico">
        
      
      
        
          <link rel="icon" type="image/png" href="/images/favicon-192x192.png" sizes="192x192">
        
      
      
        
          <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon.png">
        
      
    
    <!-- title -->
    <title>HOMER Silly Insturction</title>
    <!-- async scripts -->
    <!-- Google Analytics -->


    <!-- Umami Analytics -->


    <!-- styles -->
    
<link rel="stylesheet" href="/css/style.css">

    <!-- persian styles -->
    
    <!-- rss -->
    
    
	<!-- mathjax -->
	
<meta name="generator" content="Hexo 6.3.0"></head>

<body class="max-width mx-auto px3 ltr">
    
      <div id="header-post">
  <a id="menu-icon" href="#" aria-label="目录"><i class="fa-solid fa-bars fa-lg"></i></a>
  <a id="menu-icon-tablet" href="#" aria-label="目录"><i class="fa-solid fa-bars fa-lg"></i></a>
  <a id="top-icon-tablet" href="#" aria-label="顶部" onclick="$('html, body').animate({ scrollTop: 0 }, 'fast');" style="display:none;"><i class="fa-solid fa-chevron-up fa-lg"></i></a>
  <span id="menu">
    <span id="nav">
      <ul>
        <!--
       --><li><a href="/">首页</a></li><!--
     --><!--
       --><li><a href="/about/">关于</a></li><!--
     --><!--
       --><li><a href="/archives/">归档</a></li><!--
     --><!--
       --><li><a href="/projects/">项目</a></li><!--
     --><!--
       --><li><a href="/readtest/">CZX_plz_read_this</a></li><!--
     --><!--
       --><li><a href="/tags/">tags</a></li><!--
     --><!--
       --><li><a href="/search/">搜索</a></li><!--
     -->
      </ul>
    </span>
    <br/>
    <span id="actions">
      <ul>
        
        <li><a class="icon" aria-label="上一篇" href="/2023/06/19/test/"><i class="fa-solid fa-chevron-left" aria-hidden="true" onmouseover="$('#i-prev').toggle();" onmouseout="$('#i-prev').toggle();"></i></a></li>
        
        
        <li><a class="icon" aria-label="返回顶部" href="#" onclick="$('html, body').animate({ scrollTop: 0 }, 'fast');"><i class="fa-solid fa-chevron-up" aria-hidden="true" onmouseover="$('#i-top').toggle();" onmouseout="$('#i-top').toggle();"></i></a></li>
        <li><a class="icon" aria-label="分享文章" href="#"><i class="fa-solid fa-share-alt" aria-hidden="true" onmouseover="$('#i-share').toggle();" onmouseout="$('#i-share').toggle();" onclick="$('#share').toggle();return false;"></i></a></li>
      </ul>
      <span id="i-prev" class="info" style="display:none;">上一篇</span>
      <span id="i-next" class="info" style="display:none;">下一篇</span>
      <span id="i-top" class="info" style="display:none;">返回顶部</span>
      <span id="i-share" class="info" style="display:none;">分享文章</span>
    </span>
    <br/>
    <div id="share" style="display: none">
      <ul>
  <li><a class="icon" target="_blank" rel="noopener" href="http://www.facebook.com/sharer.php?u=http://example.com/2023/06/19/HOMER-Silly-Instruction/"><i class="fab fa-facebook " aria-hidden="true"></i></a></li>
  <li><a class="icon" target="_blank" rel="noopener" href="https://twitter.com/share?url=http://example.com/2023/06/19/HOMER-Silly-Instruction/&text=HOMER Silly Insturction"><i class="fab fa-twitter " aria-hidden="true"></i></a></li>
  <li><a class="icon" target="_blank" rel="noopener" href="http://www.linkedin.com/shareArticle?url=http://example.com/2023/06/19/HOMER-Silly-Instruction/&title=HOMER Silly Insturction"><i class="fab fa-linkedin " aria-hidden="true"></i></a></li>
  <li><a class="icon" target="_blank" rel="noopener" href="https://pinterest.com/pin/create/bookmarklet/?url=http://example.com/2023/06/19/HOMER-Silly-Instruction/&is_video=false&description=HOMER Silly Insturction"><i class="fab fa-pinterest " aria-hidden="true"></i></a></li>
  <li><a class="icon" href="mailto:?subject=HOMER Silly Insturction&body=Check out this article: http://example.com/2023/06/19/HOMER-Silly-Instruction/"><i class="fa-solid fa-envelope " aria-hidden="true"></i></a></li>
  <li><a class="icon" target="_blank" rel="noopener" href="https://getpocket.com/save?url=http://example.com/2023/06/19/HOMER-Silly-Instruction/&title=HOMER Silly Insturction"><i class="fab fa-get-pocket " aria-hidden="true"></i></a></li>
  <li><a class="icon" target="_blank" rel="noopener" href="http://reddit.com/submit?url=http://example.com/2023/06/19/HOMER-Silly-Instruction/&title=HOMER Silly Insturction"><i class="fab fa-reddit " aria-hidden="true"></i></a></li>
  <li><a class="icon" target="_blank" rel="noopener" href="http://www.stumbleupon.com/submit?url=http://example.com/2023/06/19/HOMER-Silly-Instruction/&title=HOMER Silly Insturction"><i class="fab fa-stumbleupon " aria-hidden="true"></i></a></li>
  <li><a class="icon" target="_blank" rel="noopener" href="http://digg.com/submit?url=http://example.com/2023/06/19/HOMER-Silly-Instruction/&title=HOMER Silly Insturction"><i class="fab fa-digg " aria-hidden="true"></i></a></li>
  <li><a class="icon" target="_blank" rel="noopener" href="http://www.tumblr.com/share/link?url=http://example.com/2023/06/19/HOMER-Silly-Instruction/&name=HOMER Silly Insturction&description="><i class="fab fa-tumblr " aria-hidden="true"></i></a></li>
  <li><a class="icon" target="_blank" rel="noopener" href="https://news.ycombinator.com/submitlink?u=http://example.com/2023/06/19/HOMER-Silly-Instruction/&t=HOMER Silly Insturction"><i class="fab fa-hacker-news " aria-hidden="true"></i></a></li>
</ul>

    </div>
    
    
      <div id="toc">
        <ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#HOMER-%E5%9F%BA%E6%9C%AC%E5%8A%9F%E8%83%BD"><span class="toc-number">1.</span> <span class="toc-text">HOMER 基本功能</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E4%B8%80-homer%E5%9C%A8cygwin%E7%9A%84%E5%AE%89%E8%A3%85"><span class="toc-number">1.1.</span> <span class="toc-text">一.homer在cygwin的安装</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#cygwin%E5%AE%89%E8%A3%85"><span class="toc-number">1.1.1.</span> <span class="toc-text">cygwin安装</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#%E9%97%AE%E9%A2%981"><span class="toc-number">1.1.2.</span> <span class="toc-text">!!!问题1!!!</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#%E4%B8%8B%E8%BD%BDhomer%E6%AD%A3%E5%BC%8F%E5%AE%89%E8%A3%85"><span class="toc-number">1.1.3.</span> <span class="toc-text">下载homer正式安装</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#%E5%AE%89%E8%A3%85homer%E4%B8%AD%E7%9A%84package%E6%95%B0%E6%8D%AE%E5%BA%93"><span class="toc-number">1.1.4.</span> <span class="toc-text">安装homer中的package数据库</span></a></li></ol></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E4%BA%8C-homer%E7%9A%84%E5%8A%9F%E8%83%BD%E5%92%8C%E4%BD%BF%E7%94%A8"><span class="toc-number">1.2.</span> <span class="toc-text">二.homer的功能和使用</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#1-%E5%9F%BA%E4%BA%8E%E5%9F%BA%E5%9B%A0-x2F-%E5%90%AF%E5%8A%A8%E5%AD%90%E7%9A%84motif%E5%88%86%E6%9E%90"><span class="toc-number">1.2.1.</span> <span class="toc-text">1.基于基因&#x2F;启动子的motif分析</span></a><ol class="toc-child"><li class="toc-item toc-level-5"><a class="toc-link" href="#%E4%BD%BF%E7%94%A8findMotifs-pl"><span class="toc-number">1.2.1.1.</span> <span class="toc-text">使用findMotifs.pl</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#%E8%BF%94%E5%9B%9E%E6%89%BE%E5%88%B0%E7%9A%84%E6%AF%8F%E4%B8%80%E4%B8%AAmotif%E7%9A%84%E4%BD%8D%E7%BD%AE"><span class="toc-number">1.2.1.2.</span> <span class="toc-text">返回找到的每一个motif的位置</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#%E4%BD%BF%E7%94%A8annotatePeaks-pl"><span class="toc-number">1.2.1.3.</span> <span class="toc-text">使用annotatePeaks.pl</span></a></li></ol></li><li class="toc-item toc-level-4"><a class="toc-link" href="#2-%E5%9F%BA%E4%BA%8E%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%9A%84%E5%88%86%E6%9E%90"><span class="toc-number">1.2.2.</span> <span class="toc-text">2.基于基因组的分析</span></a><ol class="toc-child"><li class="toc-item toc-level-5"><a class="toc-link" href="#%E5%9C%A8%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%8C%BA%E5%9F%9F%E4%B8%AD%E5%AF%BB%E6%89%BE%E5%AF%8C%E9%9B%86%E7%9A%84motif"><span class="toc-number">1.2.2.1.</span> <span class="toc-text">在基因组区域中寻找富集的motif</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#%E5%AF%BB%E6%89%BE%E5%85%B7%E4%BD%93motif%E7%9A%84%E4%BD%8D%E7%BD%AE"><span class="toc-number">1.2.2.2.</span> <span class="toc-text">寻找具体motif的位置</span></a></li></ol></li><li class="toc-item toc-level-4"><a class="toc-link" href="#3-%E5%9C%A8%E6%95%B4%E4%B8%AA%E5%9F%BA%E5%9B%A0%E7%BB%84%E4%B8%AD%E5%AF%BB%E6%89%BEmotif"><span class="toc-number">1.2.3.</span> <span class="toc-text">3.在整个基因组中寻找motif</span></a><ol class="toc-child"><li class="toc-item toc-level-5"><a class="toc-link" href="#%E4%BD%BF%E7%94%A8-scanMotifGenomeWide-pl"><span class="toc-number">1.2.3.1.</span> <span class="toc-text">使用 scanMotifGenomeWide.pl</span></a></li></ol></li><li class="toc-item toc-level-4"><a class="toc-link" href="#4-%E5%9C%A8mRNA%E4%B8%AD%E6%89%BE%E5%87%BARNA-motif"><span class="toc-number">1.2.4.</span> <span class="toc-text">4.在mRNA中找出RNA motif</span></a><ol class="toc-child"><li class="toc-item toc-level-5"><a class="toc-link" href="#%E5%AF%BB%E6%89%BERNA-motifs%E7%9A%84Co-regulated%E5%9F%BA%E5%9B%A0%E5%88%97%E8%A1%A8"><span class="toc-number">1.2.4.1.</span> <span class="toc-text">寻找RNA motifs的Co-regulated基因列表</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#%E5%88%86%E6%9E%90motif-%E9%93%BE%E7%89%B9%E5%BC%82%E6%80%A7%E7%9A%84genome%E5%8C%BA%E5%9F%9F"><span class="toc-number">1.2.4.2.</span> <span class="toc-text">分析motif 链特异性的genome区域</span></a></li></ol></li></ol></li></ol></li></ol>
      </div>
    
  </span>
</div>

    
    <div class="content index py4 ">
        
        <article class="post h-entry" itemscope itemtype="http://schema.org/BlogPosting">
  <header>
    
    <h1 class="posttitle p-name" itemprop="name headline">
        HOMER Silly Insturction
    </h1>



    <div class="meta">
      <span class="author p-author h-card" itemprop="author" itemscope itemtype="http://schema.org/Person">
        <span class="p-name" itemprop="name">r3zgh0st</span>
      </span>
      
    <div class="postdate">
      
        <time datetime="2023-06-19T06:11:39.000Z" class="dt-published" itemprop="datePublished">2023-06-19</time>
        
      
    </div>


      

      
    <div class="article-tag">
        <i class="fa-solid fa-tag"></i>
        <a class="p-category" href="/tags/BIOINFO/" rel="tag">BIOINFO</a>, <a class="p-category" href="/tags/test/" rel="tag">test</a>
    </div>


      </span><span class="fa fa-user-o"> </span><span  style="color:#222;"> {{ post.author }}</span> 
    </div>
  </header>
  

  <div class="content e-content" itemprop="articleBody">
    <h2 id="HOMER-基本功能"><a href="#HOMER-基本功能" class="headerlink" title="HOMER 基本功能"></a>HOMER 基本功能</h2><p>HOMER(Hypergeometric Optimization of Motif EnRichment)用于Motif identification和下一代测序分析，是一款<em>很方便</em>的软件。</p>
<p>*<em>最不方便的地方在于安装… 笔者花了整整两天才部署完，有很多小坑。</em></p>
<h3 id="一-homer在cygwin的安装"><a href="#一-homer在cygwin的安装" class="headerlink" title="一.homer在cygwin的安装"></a>一.homer在cygwin的安装</h3><p>网路上已有很多现成的安装教程资料了.这边仅记录一些在cygwin中安装的问题和解决方法.</p>
<ol>
<li><h4 id="cygwin安装"><a href="#cygwin安装" class="headerlink" title="cygwin安装"></a>cygwin安装</h4><p>下载安装好cygwin, <u>添加至系统环境变量</u>. PATH&#x3D;$PATH:&#x2F;….(安装路径)</p>
<p>注意:安装时添加好<a target="_blank" rel="noopener" href="http://mirrors.aliyun.com/cygwin/">http://mirrors.aliyun.com/cygwin/</a> 镜像网站,安装包会快一些.为了正常使用homer, 安装时添加好包:</p>
<ul>
<li><p>gcc</p>
</li>
<li><p>g++</p>
</li>
<li><p>make</p>
</li>
<li><p>perl</p>
</li>
<li><p>zip&#x2F;unzip</p>
</li>
<li><p>gzip&#x2F;gunzip</p>
</li>
<li><p>wget</p>
<p>如果一开始缺了包不要紧, 重新运行cygwin的setup程序添加就可以了, 不是重新安装！</p>
</li>
</ul>
<p>R中装好DESeq2, edgeR (一般生信用过R的都有吧)</p>
</li>
<li><h4 id="问题1"><a href="#问题1" class="headerlink" title="!!!问题1!!!"></a><strong>!!!问题1!!!</strong></h4><p>之后在用homer的时候, 发现没有motif的可视化, 查看log日志发现是缺少了seqlogo, 在python和R中装了也没用, 其实它是包含在weblogo这里的. (在homer官方安装说明中这一段已被设置灰色不再使用,但事实上还是需要使用的). <strong>所以这里还要安装几个依赖</strong>. 如果cygwin中不提供添加, 就用<code>apt-cyg install yourpackacge</code> 和以上一样全都要添加至 ~&#x2F;.bash_profile或者 ~&#x2F;.bashrc中.</p>
<ul>
<li>GMT (自带ghostscript,如果提示没安装就单独再下载一次gs)</li>
<li>ncurses-libs, ncurses-devel, xz-devel, zlib-devel, HTSlib (samtools需要)</li>
<li>samtools <a target="_blank" rel="noopener" href="http://sourceforge.net/projects/samtools/files/">http://sourceforge.net/projects/samtools/files/</a> </li>
<li>Weblogo (version 2.8.2, 不要用3 )<a target="_blank" rel="noopener" href="http://weblogo.berkeley.edu/">http://weblogo.berkeley.edu/</a>  下载后解压,不需要自己安装。</li>
<li>libpng12, apache2, uuid-dev (R), mysqldb (python)，httpd, hdf5, libuuid-devel (R), MySQL(python)  这边分debian&#x2F;ubuntu 和 CentOS, 这是blat依赖包。我用不着, 要是保险点都装吧。</li>
<li><del>blat</del> 我没有成功安装, 如果不需要处理一些专门的CHIP-seq分析也用不着。</li>
</ul>
</li>
<li><h4 id="下载homer正式安装"><a href="#下载homer正式安装" class="headerlink" title="下载homer正式安装"></a>下载homer正式安装</h4><ul>
<li>homer官网. 下载configureHomer.pl <a target="_blank" rel="noopener" href="http://homer.ucsd.edu/homer/configureHomer.pl">http://homer.ucsd.edu/homer/configureHomer.pl</a></li>
<li><code>perl /Users/chucknorris/homer/configureHomer.pl -install</code><br>由于网络问题, 一直没成功. 后来莫名其妙用<code>perl /configureHomer.pl -install homer</code>下载安装好了。</li>
</ul>
</li>
<li><h4 id="安装homer中的package数据库"><a href="#安装homer中的package数据库" class="headerlink" title="安装homer中的package数据库"></a>安装homer中的package数据库</h4><p>这边我用的是人的, 所以就以此举例.</p>
<ul>
<li>人基因组: <code>perl /configureHomer.pl -install hg38</code>  #很大,1.4G</li>
<li>人启动子: <code>perl configureHomer.pl -install human</code></li>
</ul>
<p>如果需要其他的包, 可以用 <code>perl /configureHomer.pl -list</code> 查看有哪些可以安装的.</p>
</li>
</ol>
<h3 id="二-homer的功能和使用"><a href="#二-homer的功能和使用" class="headerlink" title="二.homer的功能和使用"></a>二.homer的功能和使用</h3><h4 id="1-基于基因-x2F-启动子的motif分析"><a href="#1-基于基因-x2F-启动子的motif分析" class="headerlink" title="1.基于基因&#x2F;启动子的motif分析"></a>1.基于基因&#x2F;启动子的motif分析</h4><ol>
<li><h5 id="使用findMotifs-pl"><a href="#使用findMotifs-pl" class="headerlink" title="使用findMotifs.pl"></a>使用<code>findMotifs.pl</code></h5><p>①功能: <strong>寻找目标基因启动子中富集的motif</strong>. e.g.受某处理上调的基因, 特定细胞的基因等等.</p>
<p>②input: 一个你感兴趣的基因列表. 格式包括: </p>
<ul>
<li>NCBI Entrez Gene IDs</li>
<li>NCBI Unigene IDs</li>
<li>NCBI Refseq IDs (mRNA, protein)</li>
<li>Ensembl Gene IDs</li>
<li>Gene Symbols (i.e. Official Gene names, like “Nfkb1” )</li>
<li>popular affymetrix probe IDs (MOE430, U133plus, U95, U75A)</li>
</ul>
<p>③运行命令</p>
<p><code>findMotifs.pl &lt;inputfile.txt&gt; &lt;promoter set&gt; &lt;output directory&gt; [options]</code></p>
<p>&lt;&gt;中的内容是必须的. 其中, <promoter set>在以下种类中选择:</p>
<ul>
<li>human (<em>Homo sapiens</em>)</li>
<li>mouse (<em>Mus musculus</em>)</li>
<li>rat (<em>Rattus norvegicus</em>)</li>
<li>fly (<em>Drosophila melanogaster</em>)</li>
<li>worm (<em>Caenorhabditis elegans</em>)</li>
<li>zebrafish (<em>Danio rerio</em>)</li>
<li>yeast (<em>Saccharomyces cerevisiae</em>)</li>
</ul>
<p>※一个例子:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#目标序列在人类mRNA上富集的motif</span></span><br><span class="line">findMotifs.pl geneslist.txt human-mRNA output/ -rna -len 8</span><br></pre></td></tr></table></figure>

<p>④其他可选参数</p>
<p>请参见 <a target="_blank" rel="noopener" href="http://homer.ucsd.edu/homer/microarray/index.html">官方介绍</a> 的<em>Important motif finding parameters</em>部分:</p>
<p>包括规定masked序列, <strong>自定义启动子区域</strong>, 删除冗余启动子, 筛选<strong>motif长度</strong>, motif个数, 只在＋链寻找motif, 寻找oligos富集, <strong>转换至人ID进行GO分析</strong>, 标准化CpG%, 消除lower order oligos带来的bias, 自定义背景基因, 在全局优化中允许的错配, 选择二项式&#x2F;超几何分布打分, CPU核数。</p>
</li>
<li><h5 id="返回找到的每一个motif的位置"><a href="#返回找到的每一个motif的位置" class="headerlink" title="返回找到的每一个motif的位置"></a>返回找到的每一个motif的位置</h5><p>①在<code>findMotifs.pl</code>中这个功能不是默认的.</p>
<p>②input: 基因列表. 格式同上.</p>
<p>③使用: 加上 <code>-find &lt;motif file&gt;</code>参数. 这个文件(.motif)在初始的分析中可以得到。</p>
<p>※一个例子:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#目标序列在人类mRNA上富集的motif</span></span><br><span class="line">findMotifs.pl geneslist.txt human output/  -find motifs.motif &gt;  output.txt</span><br></pre></td></tr></table></figure>

<p>③运行结果&#x2F;输出内容:</p>
<ol>
<li>Peak&#x2F;Region ID</li>
<li>Offset from the TSS</li>
<li>Sequence of the site</li>
<li>Name of the Motif</li>
<li>Strand</li>
<li>Motif Score (log odds score of the motif matrix, higher scores are better matches)</li>
</ol>
</li>
<li><h5 id="使用annotatePeaks-pl"><a href="#使用annotatePeaks-pl" class="headerlink" title="使用annotatePeaks.pl"></a>使用<code>annotatePeaks.pl</code></h5><p>①功能: 注释peaks, 及附近基因, GO分析, 量化ChIP-Seq tags密度, 绘制图.</p>
<p>②input: HOMER peaks file, 或者BED file. 具体见这一段说明：</p>
<blockquote>
<p>HOMER peak files should have at minimum <u>5 columns</u> (separated by TABs, additional columns will be ignored):</p>
<ul>
<li>Column1: Unique Peak ID</li>
<li>Column2: chromosome</li>
<li>Column3: starting position</li>
<li>Column4: ending position</li>
<li>Column5: Strand (+&#x2F;- or 0&#x2F;1, where 0&#x3D;”+”, 1&#x3D;”-“)</li>
</ul>
<p>BED files should have at minimum <u>6 columns</u> (separated by TABs, additional columns will be ignored)</p>
<ul>
<li>Column1: chromosome</li>
<li>Column2: starting position</li>
<li>Column3: ending position</li>
<li>Column4: Unique Peak ID</li>
<li>Column5: not used</li>
<li>Column6: Strand (+&#x2F;- or 0&#x2F;1, where 0&#x3D;”+”, 1&#x3D;”-“)</li>
</ul>
<p>In theory, HOMER will accept BED files with only 4 columns (+&#x2F;- in the 4th column), and files without unique IDs, but this is NOT recommended. For one, if you don’t have unique IDs for your regions, it’s hard to go back and figure out which region contains which peak.</p>
<p>※mac中可以用homer的<code>changeNewLine.pl &lt;filename&gt;</code>转换文件格式。</p>
</blockquote>
<p>③使用：</p>
<p><code>annotatePeaks.pl &lt;peak/BED file&gt; &lt;genome&gt; [options] &gt; &lt;output file&gt;</code></p>
<p><code>annotatePeaks.pl tss &lt;promoter set&gt; -m &lt;motif file&gt; &lt;output directory&gt;</code></p>
<p>※例子:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#i.基本的注释Genomic Annotation</span></span><br><span class="line">annotatePeaks.pl peaks.txt hg38 &gt; output.txt</span><br><span class="line"><span class="comment">#ii.在TSS而非峰模式下分析转录起始位点</span></span><br><span class="line">annotatePeaks.pl tss hg38 -size -300,50 -m motifs.motif &gt; output.txt</span><br><span class="line"><span class="comment">#iii.在peaks附近寻找motif</span></span><br><span class="line">annotatePeaks.pl peaks.txt mm8 -size 200 -m ms1.motif [cebp.motif] &gt; output.txt</span><br><span class="line"><span class="comment">#iii加上-mbed motif.bed 可以生成bed文件</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#iv.在peaks1中找到最近的peaks2 ???不是很清楚使用实例???</span></span><br><span class="line">... -p &lt;peak file 1&gt; [peak file 2] [-pdist] [-pcount] [-size]</span><br></pre></td></tr></table></figure>

<p>④输出结果</p>
<p>包括: 峰值是否在TSS (default -1kb ~ +100bp), TTS(default -100 bp to +1kb), CDS Exons, 5’ UTR Exons, 3’ UTR Exons, Introns, Intergenic, *CpG Islands,  *Repeats ( *在Detailed Annotation选项中)</p>
<blockquote>
<ol>
<li>Peak ID</li>
<li>Chromosome</li>
<li>Peak start position</li>
<li>Peak end position</li>
<li>Strand</li>
<li>Peak Score</li>
<li>FDR&#x2F;Peak Focus Ratio&#x2F;Region Size</li>
<li>Annotation (i.e. Exon, Intron, …)</li>
<li>Detailed Annotation (Exon, Intron etc. + CpG Islands, repeats, etc.)</li>
<li>Distance to nearest RefSeq TSS</li>
<li>Nearest TSS: Native ID of annotation file</li>
<li>Nearest TSS: Entrez Gene ID</li>
<li>Nearest TSS: Unigene ID</li>
<li>Nearest TSS: RefSeq ID</li>
<li>Nearest TSS: Ensembl ID</li>
<li>Nearest TSS: Gene Symbol</li>
<li>Nearest TSS: Gene Aliases</li>
<li>Nearest TSS: Gene description</li>
<li>Additional columns depend on options selected when running the program.</li>
</ol>
</blockquote>
<p>⑤可选参数</p>
<ul>
<li><p><code>annotation.pl &lt;peak file&gt; &lt;genome&gt; -gene &lt;gene data file&gt;  &gt; output.txt</code></p>
<p>根据peaks最近的注释TSS, <strong>向peaks添加<u>特定基因</u>的信息</strong>。</p>
</li>
<li><p>有关diagram plot 详见: <a target="_blank" rel="noopener" href="http://homer.ucsd.edu/homer/ngs/quantification.html">ann可视化</a></p>
<p>※一些例子. 用excel打开, 作图 (除了热图)</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">###scatter</span></span><br><span class="line"><span class="comment">#me1和me3在小鼠胚胎干细胞峰peaks附近的分布 *log--&gt; -log2</span></span><br><span class="line">annotatePeaks.pl peaks.txt mm8 -size 1000 -d H3K4me1-ChIP-Seq/ H3K4me3-ChIP-Seq/ &gt; output.txt</span><br><span class="line"></span><br><span class="line"><span class="comment">###hist</span></span><br><span class="line"><span class="comment">#TSS相关的Motif YY1的分布</span></span><br><span class="line">annotatePeaks.pl tss mm9 -size -500,250 -hist 10 -m yy1.motif &gt; output.txt</span><br><span class="line"></span><br><span class="line"><span class="comment">###positions ???目录下的tags是什么类型的文件??? 5‘tag</span></span><br><span class="line"><span class="comment">#motifs near peaks</span></span><br><span class="line">annotatePeaks.pl peaks.txt hg18 -size 6000 -hist 25 -m &lt;are.motif&gt; [fox.motif ap1.motif]</span><br><span class="line"><span class="comment">#genes/tags near peaks</span></span><br><span class="line">annotatePeaks.pl &lt;peak file&gt; &lt;genome&gt; -size &lt;<span class="comment">#&gt; -hist &lt;#&gt; -d &lt;tag directory 1&gt; [tag directory2] ... -m &lt;motif 1&gt; &lt;motif 2&gt; ... &gt;  &lt;output matrix file&gt;</span></span><br><span class="line">annotatePeaks.pl peaks.txt hg18 -size 6000 -hist 25 -d me1/ me2/ Mme3/ &gt; output.txt</span><br><span class="line"></span><br><span class="line"><span class="comment">###heatmap --&gt; matrix --&gt; visualized with other software</span></span><br><span class="line">annotatePeaks.pl &lt;peak file&gt; &lt;genome&gt; -size &lt;..&gt; -hist &lt;..&gt; -ghist -d &lt;tag directory 1&gt; [tag directory2] ... &gt; &lt;output matrix file&gt;</span><br></pre></td></tr></table></figure></li>
</ul>
<p>TSS相关的Motif YY1的分布例图:</p>
<p><img src="S:\BlogProject\blog\source_posts\HOMER-Silly-Tutorials_pics\motifs.tssHistogram.png" alt="Motif TSS histogram YY1"></p>
</li>
</ol>
<h4 id="2-基于基因组的分析"><a href="#2-基于基因组的分析" class="headerlink" title="2.基于基因组的分析"></a>2.基于基因组的分析</h4><ol>
<li><h5 id="在基因组区域中寻找富集的motif"><a href="#在基因组区域中寻找富集的motif" class="headerlink" title="在基因组区域中寻找富集的motif"></a>在基因组区域中寻找富集的motif</h5><p>①input: HOMER peak file, BED file, 具体格式见前文引用。</p>
<p>②使用:</p>
<p><code>findMotifsGenome.pl &lt;peak/BED file&gt; &lt;genome&gt; &lt;output directory&gt; -size # [options]</code></p>
<p>③输出结果:</p>
<ul>
<li><p>homerMotifs.motifs&lt;#&gt; :从头搜寻的motif, 信息有长度, 算法单独运行</p>
</li>
<li><p>homerMotifs.all.motifs : 全部的motifs</p>
</li>
<li><p>motifFindingParameters.txt : 键入的命令</p>
</li>
<li><p>knownResults.txt : 统计数据, 可在excel打开</p>
</li>
<li><p>seq.autonorm.tsv : lower-order oligo的自动标准化</p>
</li>
<li><p>homerResults.html : 从头搜寻motif的格式化</p>
</li>
<li><p>homerResults&#x2F; 上面的内容</p>
</li>
<li><p>knownResults.html: 已知的motif输出</p>
</li>
<li><p>knownResults&#x2F; 上面的内容</p>
</li>
</ul>
<p>④可选参数详见:  <a target="_blank" rel="noopener" href="http://homer.ucsd.edu/homer/ngs/peakMotifs.html">findMotifsGenome</a></p>
<ul>
<li><p>背景, motif长度, 用于motif搜寻的区域长度, 不搜索de novo的motif, 等等</p>
</li>
<li><p>重要的一个: <strong>-rna</strong> 用RNA数据, 输出mRNA motifs, 关于没有其他RNA数据库, 作者写了一句很可爱的话:</p>
<blockquote>
<p> I guess chuck roundhouse kicked all of the splicing and other RNA motifs into hard to find databases.</p>
</blockquote>
</li>
</ul>
</li>
<li><h5 id="寻找具体motif的位置"><a href="#寻找具体motif的位置" class="headerlink" title="寻找具体motif的位置"></a>寻找具体motif的位置</h5><p>①在<code>findMotifsGenome.pl</code>中这个功能也不是默认的.</p>
<p>②input： 从使用homer得到筛选你感兴趣的motifs之后, 使用 <code>-find &lt;motifile&gt;</code></p>
<p>③使用:</p>
<p>i.<code>findMotifsGenome.pl &lt;peaks.txt/bed&gt; &lt;genome&gt; /out -find &lt;motif file&gt; /output.txt</code></p>
<p>ii.<code>annotatePeaks.pl  &lt;peaks&gt; &lt;genome&gt; -m motifs.motif &gt; output.txt</code></p>
<p>④输出结果</p>
<p>在i中:</p>
<blockquote>
<ol>
<li>Peak&#x2F;Region ID</li>
<li>Offset from the center of the region</li>
<li>Sequence of the site</li>
<li>Name of the Motif</li>
<li>Strand</li>
<li>Motif Score (log odds score of the motif matrix, higher scores are better matches)</li>
</ol>
</blockquote>
<p>在ii中:</p>
<blockquote>
<ol>
<li>Peak&#x2F;Region ID</li>
<li>Chromosome</li>
<li>Start</li>
<li>End</li>
<li>Strand of Peaks</li>
</ol>
<p>  6-18: annotation information</p>
<ol start="19">
<li>CpG%</li>
<li>GC%</li>
<li>Motif Instances<br>  …</li>
</ol>
</blockquote>
<p>⑤可视化</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#motif密度 -hist &lt;..&gt;</span></span><br><span class="line"></span><br><span class="line">annotatePeaks.pl peaks.txt hg38 -m m1.motif m2.motif -size 1000 -hist 10 &gt; output.txt</span><br></pre></td></tr></table></figure></li>
</ol>
<h4 id="3-在整个基因组中寻找motif"><a href="#3-在整个基因组中寻找motif" class="headerlink" title="3.在整个基因组中寻找motif"></a>3.在整个基因组中寻找motif</h4><ol>
<li><h5 id="使用-scanMotifGenomeWide-pl"><a href="#使用-scanMotifGenomeWide-pl" class="headerlink" title="使用 scanMotifGenomeWide.pl"></a>使用 <code>scanMotifGenomeWide.pl</code></h5><p>①input: motif file</p>
<p>②使用: </p>
<p><code>scanMotifGenomeWide.pl &lt;motif file&gt; &lt;genome&gt; [options]</code></p>
<p>※一个例子:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#输出为bed, 默认txt</span></span><br><span class="line">scanMotifGenomeWide.pl pu1.motif mm9 -bed &gt; pu1.sites.mm9.bed</span><br></pre></td></tr></table></figure>

<p>③输出结果:</p>
<blockquote>
<p>Tab delimited text file (default):</p>
<ol>
<li>Site ID (motif name + number)</li>
<li>chr</li>
<li>start</li>
<li>end</li>
<li>strand</li>
<li>log-odds score</li>
<li>sequence</li>
</ol>
<p>BED (tab) format (use <strong>-bed</strong>):</p>
<ol>
<li>chr</li>
<li>start</li>
<li>end</li>
<li>motif name</li>
<li>log-odds score (will be floored to an integer)</li>
<li>strand</li>
</ol>
</blockquote>
<p>④可选参数详见<a target="_blank" rel="noopener" href="http://homer.ucsd.edu/homer/motif/genomeWideMotifScan.html">genenomeWide</a></p>
</li>
</ol>
<h4 id="4-在mRNA中找出RNA-motif"><a href="#4-在mRNA中找出RNA-motif" class="headerlink" title="4.在mRNA中找出RNA motif"></a>4.在mRNA中找出RNA motif</h4><p>基本的使用:在2.1已经提到了, 在<code>findMotifs.pl</code>和<code> and findMotifsGenome.pl</code>中加入<code>-rna</code>即可.</p>
<ol>
<li><h5 id="寻找RNA-motifs的Co-regulated基因列表"><a href="#寻找RNA-motifs的Co-regulated基因列表" class="headerlink" title="寻找RNA motifs的Co-regulated基因列表"></a>寻找RNA motifs的Co-regulated基因列表</h5><p>①使用</p>
<p>※一个例子</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">findMotifs.pl mdownregulated.genes.txt human-mRNA Output/ -rna -len 8</span><br></pre></td></tr></table></figure>

<p>其实规定了human-mRNA就不用再加<code>-rna</code>了。</p>
<p>现在也尝试match人类的miRNA seeds (miRBase)</p>
<p>②可选参数</p>
<p>在<code>findMotifs.pl</code>中可以添加:</p>
<ul>
<li>-min &lt;#&gt;:要考虑的最小mRNA长度 (移除极短的mRNA序列)</li>
<li>-max &lt;#&gt;:要考虑的最大mRNA长度 (删除非常长的RNA)</li>
</ul>
</li>
<li><h5 id="分析motif-链特异性的genome区域"><a href="#分析motif-链特异性的genome区域" class="headerlink" title="分析motif 链特异性的genome区域"></a>分析motif 链特异性的genome区域</h5><p><code>findMotifsGenome.pl fox2.clip.bed hg17 MotifOutput -rna</code></p>
</li>
</ol>

  </div>
</article>



        
          <div id="footer-post-container">
  <div id="footer-post">

    <div id="nav-footer" style="display: none">
      <ul>
        
          <li><a href="/">首页</a></li>
        
          <li><a href="/about/">关于</a></li>
        
          <li><a href="/archives/">归档</a></li>
        
          <li><a href="/projects/">项目</a></li>
        
          <li><a href="/readtest/">CZX_plz_read_this</a></li>
        
          <li><a href="/tags/">tags</a></li>
        
          <li><a href="/search/">搜索</a></li>
        
      </ul>
    </div>

    
    
      <div id="toc-footer" style="display: none">
        <ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#HOMER-%E5%9F%BA%E6%9C%AC%E5%8A%9F%E8%83%BD"><span class="toc-number">1.</span> <span class="toc-text">HOMER 基本功能</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E4%B8%80-homer%E5%9C%A8cygwin%E7%9A%84%E5%AE%89%E8%A3%85"><span class="toc-number">1.1.</span> <span class="toc-text">一.homer在cygwin的安装</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#cygwin%E5%AE%89%E8%A3%85"><span class="toc-number">1.1.1.</span> <span class="toc-text">cygwin安装</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#%E9%97%AE%E9%A2%981"><span class="toc-number">1.1.2.</span> <span class="toc-text">!!!问题1!!!</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#%E4%B8%8B%E8%BD%BDhomer%E6%AD%A3%E5%BC%8F%E5%AE%89%E8%A3%85"><span class="toc-number">1.1.3.</span> <span class="toc-text">下载homer正式安装</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#%E5%AE%89%E8%A3%85homer%E4%B8%AD%E7%9A%84package%E6%95%B0%E6%8D%AE%E5%BA%93"><span class="toc-number">1.1.4.</span> <span class="toc-text">安装homer中的package数据库</span></a></li></ol></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E4%BA%8C-homer%E7%9A%84%E5%8A%9F%E8%83%BD%E5%92%8C%E4%BD%BF%E7%94%A8"><span class="toc-number">1.2.</span> <span class="toc-text">二.homer的功能和使用</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#1-%E5%9F%BA%E4%BA%8E%E5%9F%BA%E5%9B%A0-x2F-%E5%90%AF%E5%8A%A8%E5%AD%90%E7%9A%84motif%E5%88%86%E6%9E%90"><span class="toc-number">1.2.1.</span> <span class="toc-text">1.基于基因&#x2F;启动子的motif分析</span></a><ol class="toc-child"><li class="toc-item toc-level-5"><a class="toc-link" href="#%E4%BD%BF%E7%94%A8findMotifs-pl"><span class="toc-number">1.2.1.1.</span> <span class="toc-text">使用findMotifs.pl</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#%E8%BF%94%E5%9B%9E%E6%89%BE%E5%88%B0%E7%9A%84%E6%AF%8F%E4%B8%80%E4%B8%AAmotif%E7%9A%84%E4%BD%8D%E7%BD%AE"><span class="toc-number">1.2.1.2.</span> <span class="toc-text">返回找到的每一个motif的位置</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#%E4%BD%BF%E7%94%A8annotatePeaks-pl"><span class="toc-number">1.2.1.3.</span> <span class="toc-text">使用annotatePeaks.pl</span></a></li></ol></li><li class="toc-item toc-level-4"><a class="toc-link" href="#2-%E5%9F%BA%E4%BA%8E%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%9A%84%E5%88%86%E6%9E%90"><span class="toc-number">1.2.2.</span> <span class="toc-text">2.基于基因组的分析</span></a><ol class="toc-child"><li class="toc-item toc-level-5"><a class="toc-link" href="#%E5%9C%A8%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%8C%BA%E5%9F%9F%E4%B8%AD%E5%AF%BB%E6%89%BE%E5%AF%8C%E9%9B%86%E7%9A%84motif"><span class="toc-number">1.2.2.1.</span> <span class="toc-text">在基因组区域中寻找富集的motif</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#%E5%AF%BB%E6%89%BE%E5%85%B7%E4%BD%93motif%E7%9A%84%E4%BD%8D%E7%BD%AE"><span class="toc-number">1.2.2.2.</span> <span class="toc-text">寻找具体motif的位置</span></a></li></ol></li><li class="toc-item toc-level-4"><a class="toc-link" href="#3-%E5%9C%A8%E6%95%B4%E4%B8%AA%E5%9F%BA%E5%9B%A0%E7%BB%84%E4%B8%AD%E5%AF%BB%E6%89%BEmotif"><span class="toc-number">1.2.3.</span> <span class="toc-text">3.在整个基因组中寻找motif</span></a><ol class="toc-child"><li class="toc-item toc-level-5"><a class="toc-link" href="#%E4%BD%BF%E7%94%A8-scanMotifGenomeWide-pl"><span class="toc-number">1.2.3.1.</span> <span class="toc-text">使用 scanMotifGenomeWide.pl</span></a></li></ol></li><li class="toc-item toc-level-4"><a class="toc-link" href="#4-%E5%9C%A8mRNA%E4%B8%AD%E6%89%BE%E5%87%BARNA-motif"><span class="toc-number">1.2.4.</span> <span class="toc-text">4.在mRNA中找出RNA motif</span></a><ol class="toc-child"><li class="toc-item toc-level-5"><a class="toc-link" href="#%E5%AF%BB%E6%89%BERNA-motifs%E7%9A%84Co-regulated%E5%9F%BA%E5%9B%A0%E5%88%97%E8%A1%A8"><span class="toc-number">1.2.4.1.</span> <span class="toc-text">寻找RNA motifs的Co-regulated基因列表</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#%E5%88%86%E6%9E%90motif-%E9%93%BE%E7%89%B9%E5%BC%82%E6%80%A7%E7%9A%84genome%E5%8C%BA%E5%9F%9F"><span class="toc-number">1.2.4.2.</span> <span class="toc-text">分析motif 链特异性的genome区域</span></a></li></ol></li></ol></li></ol></li></ol>
      </div>
    

    <div id="share-footer" style="display: none">
      <ul>
  <li><a class="icon" target="_blank" rel="noopener" href="http://www.facebook.com/sharer.php?u=http://example.com/2023/06/19/HOMER-Silly-Instruction/"><i class="fab fa-facebook fa-lg" aria-hidden="true"></i></a></li>
  <li><a class="icon" target="_blank" rel="noopener" href="https://twitter.com/share?url=http://example.com/2023/06/19/HOMER-Silly-Instruction/&text=HOMER Silly Insturction"><i class="fab fa-twitter fa-lg" aria-hidden="true"></i></a></li>
  <li><a class="icon" target="_blank" rel="noopener" href="http://www.linkedin.com/shareArticle?url=http://example.com/2023/06/19/HOMER-Silly-Instruction/&title=HOMER Silly Insturction"><i class="fab fa-linkedin fa-lg" aria-hidden="true"></i></a></li>
  <li><a class="icon" target="_blank" rel="noopener" href="https://pinterest.com/pin/create/bookmarklet/?url=http://example.com/2023/06/19/HOMER-Silly-Instruction/&is_video=false&description=HOMER Silly Insturction"><i class="fab fa-pinterest fa-lg" aria-hidden="true"></i></a></li>
  <li><a class="icon" href="mailto:?subject=HOMER Silly Insturction&body=Check out this article: http://example.com/2023/06/19/HOMER-Silly-Instruction/"><i class="fa-solid fa-envelope fa-lg" aria-hidden="true"></i></a></li>
  <li><a class="icon" target="_blank" rel="noopener" href="https://getpocket.com/save?url=http://example.com/2023/06/19/HOMER-Silly-Instruction/&title=HOMER Silly Insturction"><i class="fab fa-get-pocket fa-lg" aria-hidden="true"></i></a></li>
  <li><a class="icon" target="_blank" rel="noopener" href="http://reddit.com/submit?url=http://example.com/2023/06/19/HOMER-Silly-Instruction/&title=HOMER Silly Insturction"><i class="fab fa-reddit fa-lg" aria-hidden="true"></i></a></li>
  <li><a class="icon" target="_blank" rel="noopener" href="http://www.stumbleupon.com/submit?url=http://example.com/2023/06/19/HOMER-Silly-Instruction/&title=HOMER Silly Insturction"><i class="fab fa-stumbleupon fa-lg" aria-hidden="true"></i></a></li>
  <li><a class="icon" target="_blank" rel="noopener" href="http://digg.com/submit?url=http://example.com/2023/06/19/HOMER-Silly-Instruction/&title=HOMER Silly Insturction"><i class="fab fa-digg fa-lg" aria-hidden="true"></i></a></li>
  <li><a class="icon" target="_blank" rel="noopener" href="http://www.tumblr.com/share/link?url=http://example.com/2023/06/19/HOMER-Silly-Instruction/&name=HOMER Silly Insturction&description="><i class="fab fa-tumblr fa-lg" aria-hidden="true"></i></a></li>
  <li><a class="icon" target="_blank" rel="noopener" href="https://news.ycombinator.com/submitlink?u=http://example.com/2023/06/19/HOMER-Silly-Instruction/&t=HOMER Silly Insturction"><i class="fab fa-hacker-news fa-lg" aria-hidden="true"></i></a></li>
</ul>

    </div>

    <div id="actions-footer">
        <a id="menu" class="icon" href="#" onclick="$('#nav-footer').toggle();return false;"><i class="fa-solid fa-bars fa-lg" aria-hidden="true"></i> 菜单</a>
        
          <a id="toc" class="icon" href="#" onclick="$('#toc-footer').toggle();return false;"><i class="fa-solid fa-list fa-lg" aria-hidden="true"></i> 目录</a>
        
        <a id="share" class="icon" href="#" onclick="$('#share-footer').toggle();return false;"><i class="fa-solid fa-share-alt fa-lg" aria-hidden="true"></i> 分享</a>
        <a id="top" style="display:none" class="icon" href="#" onclick="$('html, body').animate({ scrollTop: 0 }, 'fast');"><i class="fa-solid fa-chevron-up fa-lg" aria-hidden="true"></i> 返回顶部</a>
    </div>

  </div>
</div>

        
        <footer id="footer">
  <div class="footer-left">
    Copyright &copy;
    
    
    2023
    r3zgh0st
  </div>
  <div class="footer-right">
    <nav>
      <ul>
        <!--
       --><li><a href="/">首页</a></li><!--
     --><!--
       --><li><a href="/about/">关于</a></li><!--
     --><!--
       --><li><a href="/archives/">归档</a></li><!--
     --><!--
       --><li><a href="/projects/">项目</a></li><!--
     --><!--
       --><li><a href="/readtest/">CZX_plz_read_this</a></li><!--
     --><!--
       --><li><a href="/tags/">tags</a></li><!--
     --><!--
       --><li><a href="/search/">搜索</a></li><!--
     -->
      </ul>
    </nav>
  </div>
</footer>

    </div>
    <!-- styles -->



  <link rel="preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.0/css/all.min.css" crossorigin="anonymous" onload="this.onload=null;this.rel='stylesheet'"/>


    <!-- jquery -->

  <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.6.0/jquery.min.js" crossorigin="anonymous"></script>




<!-- clipboard -->

  
    <script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.7/clipboard.min.js" crossorigin="anonymous"></script>
  
  <script type="text/javascript">
  $(function() {
    // copy-btn HTML
    var btn = "<span class=\"btn-copy tooltipped tooltipped-sw\" aria-label=\"复制到粘贴板！\">";
    btn += '<i class="fa-regular fa-clone"></i>';
    btn += '</span>';
    // mount it!
    $(".highlight table").before(btn);
    var clip = new ClipboardJS('.btn-copy', {
      text: function(trigger) {
        return Array.from(trigger.nextElementSibling.querySelectorAll('.code')).reduce((str,it)=>str+it.innerText+'\n','')
      }
    });
    clip.on('success', function(e) {
      e.trigger.setAttribute('aria-label', "复制成功！");
      e.clearSelection();
    })
  })
  </script>


<script src="/js/main.js"></script>

<!-- search -->

<!-- Baidu Analytics -->

<!-- Cloudflare Analytics -->

<!-- Disqus Comments -->

<!-- utterances Comments -->

</body>
</html>
